### PR TITLE
 GTC-2824 Use raster GADM layers for pro dashboard for small # of features

### PR DESCRIPTION
**Pull request by danscales**
GFWPro Dashboard currently reads in the entire vector GADM file (1.9GB) in order to do a spatial join with the user input features to determine the GADM areas in the input locations and dissolved lists.

Both reading the the GADM file and doing the spatial join can take many minutes, even in the case of a small number of input features. Instead, in that case, we change to using the raster GADM layers to determine the relevant gadm locations for the locations and dissolved lists. Because we lazily read tiles for the GFWPro dashboard (see following commit), we don't read any of the GADM rasters in the case that we are using the vector GADM.

In performance runs, I have seen current batch jobs with 20-50 features that run in 6-9 minutes, but take only about 1 minutes with the same batch configuration when using the raster GADM approach. This should be useful for reducing the cost of the hundreds of batch jobs used to update for the dashboards for all the smaller customer lists every week on Saturday night.

### CODE REVIEW

File: src/main/scala/org/globalforestwatch/summarystats/gfwpro_dashboard/GfwProDashboardAnalysis.scala [hunk 1]
```diff
@@ -23,32 +23,72 @@ object GfwProDashboardAnalysis extends SummaryAnalysis {
...
+         * The resulting FeatureId carries combined identity of source feature and contextual geometry
+         */
         SpatialJoinRDD
           .flatSpatialJoin(spatialContextualRDD, spatialFeatureRDD, considerBoundaryIntersection = true, usingIndex = true)
           .rdd
           .flatMap { case (feature, context) =>
-            refineContextualIntersection(feature, context, contextualFeatureType)
+            refineContextualIntersection(feature, context, "gadm")
+          }
+      } else {
```
Comment (jterry64) [inline src/main/scala/org/globalforestwatch/summarystats/gfwpro_dashboard/GfwProDashboardAnalysis.scala:61]: Haven't you already called the method to split into 1x1 grid by this point, so potentially the centroid is actually the centroid of a tile?

File: src/main/scala/org/globalforestwatch/summarystats/gfwpro_dashboard/GfwProDashboardAnalysis.scala [hunk 2]
```diff
@@ -23,32 +23,72 @@ object GfwProDashboardAnalysis extends SummaryAnalysis {
...
+        println("Using raster gadm")
+        rdd.map {
```
Comment (jterry64) [inline src/main/scala/org/globalforestwatch/summarystats/gfwpro_dashboard/GfwProDashboardAnalysis.scala:63]: I can't seem to find where this `rdd` is defined, is it the `featureRDD`?

File: src/main/scala/org/globalforestwatch/summarystats/gfwpro_dashboard/GfwProDashboardGridSources.scala [hunk 3]
```diff
@@ -14,30 +13,19 @@ case class GfwProDashboardGridSources(gridTile: GridTile, kwargs: Map[String, An
...
-        .catchNonFatal(sbtnNaturalForest.fetchWindow(windowKey, windowLayout))
-        .right
-      jrcForestCoverTile <- Either
-        .catchNonFatal(jrcForestCover.fetchWindow(windowKey, windowLayout))
-        .right
-    } yield {
-      val tile = GfwProDashboardTile(integratedAlertsTile, tcd2000Tile, sbtnNaturalForestTile, jrcForestCoverTile)
-      Raster(tile, windowKey.extent(windowLayout))
-    }
+    val tile = GfwProDashboardTile(
```
Comment (jterry64) [inline src/main/scala/org/globalforestwatch/summarystats/gfwpro_dashboard/GfwProDashboardGridSources.scala:25]: why this change?

File: src/main/scala/org/globalforestwatch/summarystats/gfwpro_dashboard/GfwProDashboardSummary.scala [hunk 4]
```diff
@@ -26,11 +26,23 @@ case class GfwProDashboardSummary(
   def isEmpty = stats.isEmpty
 
   /** Pivot raw data to GfwProDashboardData and aggregate across alert dates. */
-  def toGfwProDashboardData(): GfwProDashboardData = {
-    stats
-      .map { case (group, data) => group.
-        toGfwProDashboardData(data.alertCount, data.treeCoverExtentArea) }
-      .foldLeft(GfwProDashboardData.empty)( _ merge _)
+  def toGfwProDashboardData(ignoreGadm: Boolean): List[GfwProDashboardData] = {
+    if (ignoreGadm) {
```
Comment (jterry64) [inline src/main/scala/org/globalforestwatch/summarystats/gfwpro_dashboard/GfwProDashboardSummary.scala:30]: This works, but I feel like having the split between raster for small number of features and vector for big number of features adds a lot of complication/splitting paths to the code. 

Is there a good way to do raster for all cases eventually? I think mapping polygons to centroids from an RDD is pretty fast, you could potentially pull those all out into a separate DF, repartition those points by spatial keys, and then get the Gadm ID for each. But I guess that's also fairly complicated.

File: src/main/scala/org/globalforestwatch/summarystats/gfwpro_dashboard/GfwProDashboardAnalysis.scala (hunk 1 above)
Comment (danscales) [reply src/main/scala/org/globalforestwatch/summarystats/gfwpro_dashboard/GfwProDashboardAnalysis.scala:61]: Actually, we were computing the centroid of the tile after the 1x1 split since the original Dashboard code long ago, but I recently fixed this long-standing bug with PR #236 (gtc-2859, gpv-3182).  Now, we compute the centroid of the entire location BEFORE the split and stash it in the FeatureID of each tile.  See the comment in refineContextualIntersection() at line 159.

File: src/main/scala/org/globalforestwatch/summarystats/gfwpro_dashboard/GfwProDashboardAnalysis.scala (hunk 2 above)
Comment (danscales) [reply src/main/scala/org/globalforestwatch/summarystats/gfwpro_dashboard/GfwProDashboardAnalysis.scala:63]: This rdd is defined at line 45 - it's the argument of the main flatMap call.  So, its the elements of featureRDD that are current valid (have not attached problem/error already).

File: src/main/scala/org/globalforestwatch/summarystats/gfwpro_dashboard/GfwProDashboardGridSources.scala (hunk 3 above)
Comment (danscales) [reply src/main/scala/org/globalforestwatch/summarystats/gfwpro_dashboard/GfwProDashboardGridSources.scala:25]: This is part of making the GFWProDashboard sources be lazy.  With the lazy changes (as I already did a while ago for FCD), a tile is not fetched until it is first actually needed at run-time by the GridVisitor for a specific location.  I needed to make GFWProDashboard sources be lazy like FCD, because we don't want to load any GADM raster tiles during GridVisitor unless we are specifically using raster GADM because the # of locations is small.  I.e. If we are using the vector GADM and doing the full intersect, we don't want to drag down its performance by also loading raster gadm tiles that we won't actually use.

File: src/main/scala/org/globalforestwatch/summarystats/gfwpro_dashboard/GfwProDashboardSummary.scala (hunk 4 above)
Comment (danscales) [reply src/main/scala/org/globalforestwatch/summarystats/gfwpro_dashboard/GfwProDashboardSummary.scala:30]: Yes, I think your suggested way would add some  complexity too.  But the bigger problem is that I found that the vector GADM method is more efficient for very large dataset (100s of locations).  I think it's just that with 100s of locations, the vector intersection method eventually becomes more efficient than the raster method.  This was even true when I tried some extra complexities to compute the GADMs of the centroids of the locations during the GridVisitor (so tiles with centroids were not loaded a second time).

I do realize this optimization adds complexity, but I'm really hoping we get a big win for all the hundreds of small jobs on Saturday night, which can run in less than a minute with the raster GADM method, but take 5-8 minutes at minimum for the vector GADM method, because they have to load the entire vector GADM file.

**Review by jterry64 — APPROVED** [review]
